### PR TITLE
Add Pangeo, BeeGFS ui-meta

### PR DESCRIPTION
**Pull request by brtkwr**
Review comments addressed in https://github.com/cedadev/jasmin-appliances/pull/43

### CODE REVIEW

File: group_vars/openstack/gluster [hunk 1]
```diff
@@ -17,7 +17,7 @@ gluster_group_servers:
   flavor: "j1.small"
   image: "centos-7-20190104"
   user: "centos"
-  num_nodes: 3
+  num_nodes: "{{ gluster_num_nodes }}"
```
Comment (oneswig) [inline group_vars/openstack/gluster:20]: Should we define a value for `gluster_num_nodes`, or for consistency should we remove the value set for `beegfs_oss_num_nodes` in `group_vars/openstack/beegfs`

File: ui-meta/beegfs.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    kind: "integer"
+    options:
+      min: 2
+
+  - name: "beegfs_oss_volume_size"
+    label: "Storage volume size per node"
+    description: "BeeGFS object storage server volume size."
+    kind: "integer"
+    options:
+      min: 10
```
Comment (oneswig) [inline ui-meta/beegfs.yml:18]: I'm sure you could make a smaller one, if you wanted...

File: ui-meta/slurm.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+name: "slurm"
+label: "Slurm"
+description: "Slurm OpenHPC cluster."
+
+parameters:
```
Comment (oneswig) [inline ui-meta/slurm.yml:5]: We also need to expose user definitions, somehow.

**Review by oneswig — COMMENTED** [review]
Nice work, some minor questions.

File: group_vars/openstack/beegfs [hunk 4]
```diff
@@ -28,7 +30,7 @@ beegfs_group_oss:
   flavor: "j2.small"
   image: "centos-7-20190104"
   user: "centos"
-  num_nodes: "{{ cluster_num_workers }}"
+  num_nodes: "{{ beegfs_oss_num_nodes }}"
```
Comment (mkjpryor-stfc) [inline group_vars/openstack/beegfs:33]: As we discussed, `cluster_num_workers` isn't part of the spec so this change isn't technically required.

It is arguably more readable with `beegfs_oss_num_nodes`.

File: group_vars/openstack/gluster (hunk 1 above)
Comment (mkjpryor-stfc) [reply group_vars/openstack/gluster:20]: @oneswig Have we tested resizing a gluster cluster? If not, then we should fix it at three.

Or we could just default it to that in the ui-meta. Does gluster have the requirement that the number of nodes is odd and > 3, like a lot of HA stuff (e.g. etcd).

I don't think I have a problem with fixed-size storage clusters.

File: group_vars/openstack/beegfs (hunk 4 above)
Comment (mkjpryor-stfc) [reply group_vars/openstack/beegfs:33]: You need to either revert this change or change the name in `ui-meta/beegfs.yml`.

File: ui-meta/slurm.yml (hunk 3 above)
Comment (mkjpryor-stfc) [reply ui-meta/slurm.yml:5]: See discussion on the Slack channel

File: ui-meta/slurm.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  - name: "cluster_num_workers"
+    label: "Worker nodes"
+    description: "The number of worker nodes in the cluster."
+    kind: "integer"
+    options:
+      min: 1
+
+  - name: "storage_name"
+    label: "Storage service name"
+    description: "Name of the storage server attach to this cluster"
```
Comment (mkjpryor-stfc) [inline ui-meta/slurm.yml:15]: We should stick with the cluster nomenclature here, and be more descriptive about what it is being used for, i.e. "Home directory storage" and "Storage cluster to use for user home directories."

**Review by mkjpryor-stfc — CHANGES_REQUESTED** [review]

File: ui-meta/beegfs.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+name: "beegfs"
+label: "BeeGFS"
+description: "BeeGFS storage cluster."
+
+parameters:
+  - name: "beegfs_mds_volume_size"
+    label: "Metadata volume size"
```
Comment (mkjpryor-stfc) [inline ui-meta/beegfs.yml:7]: Maybe add a unit, i.e. "Metadata volume size (GB)"

File: ui-meta/beegfs.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    description: "BeeGFS metadata server volume size."
+    kind: "integer"
+    options:
+      min: 2
+
+  - name: "beegfs_oss_volume_size"
+    label: "Storage volume size per node"
```
Comment (mkjpryor-stfc) [inline ui-meta/beegfs.yml:14]: Maybe add a unit, i.e. "... (GB)"

Comment (mkjpryor-stfc) [reply ui-meta/beegfs.yml:14]: The per-node bit can probably live in the description, i.e. "Storage volume size" and "Size of the BeeGFS object storage volume on each node."

File: ui-meta/gluster.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+name: "gluster"
+label: "Gluster"
+description: "Gluster storage cluster."
+
+parameters:
+  - name: "gluster_volume_size"
+    label: "Volume size"
+    description: "Gluster storage server volume size per node."
```
Comment (mkjpryor-stfc) [inline ui-meta/gluster.yml:8]: Maybe add a unit, i.e. "... (GB)"

File: ui-meta/nfs.yml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+name: "nfs"
+label: "NFS"
+description: "NFS storage server."
+
+parameters:
+  - name: "nfs_volume_size"
+    label: "Volume size"
```
Comment (mkjpryor-stfc) [inline ui-meta/nfs.yml:7]: Maybe add a unit, i.e. "... (GB)"

File: ui-meta/pangeo.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    label: "Root volume size (GB)"
+    description: "The size of the root volume of cluster nodes."
+    kind: "integer"
+    options:
+      min: 10
+    immutable: true
+    default: 40
+
+  - name: "pangeo_username"
+    label: "Pangeo cluster initial user with admin privileges"
```
Comment (mkjpryor-stfc) [inline ui-meta/pangeo.yml:48]: This is a bit long. Probably "Admin username" is sufficient - they already know it is Pangeo.

File: ui-meta/pangeo.yml [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    description: "Username of the Pangeo cluster adminstrator."
+    kind: "string"
+    default: "admin"
+    
+  - name: "pangeo_password"
+    label: "Pangeo cluster password"
```
Comment (mkjpryor-stfc) [inline ui-meta/pangeo.yml:54]: Again, probably "Cluster password" is sufficient.

File: ui-meta/pangeo.yml [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    description: "Default password for the Pangeo cluster for all users."
+    kind: "string"
+    default: "secretnomore"
```
Comment (mkjpryor-stfc) [inline ui-meta/pangeo.yml:57]: I will add some options to the string type so that this can use a secret field.

**Review by mkjpryor-stfc — CHANGES_REQUESTED** [review]

File: group_vars/openstack/gluster (hunk 1 above)
Comment (brtkwr) [reply group_vars/openstack/gluster:20]: I was able to add another node to an existing gluster cluster but it reported failure under the task `TASK [stackhpc.gluster-cluster : Create a trusted storage pool]`:
```
fatal: [bharat-glusterfs-server-2]: FAILED! => {"changed": false, "msg": "", "rc": 1}
```

Comment (brtkwr) [reply group_vars/openstack/gluster:20]: I have set the min and max nodes to 3 until we manage to fix the scaling which I can look into now.

Comment (brtkwr) [reply group_vars/openstack/gluster:20]: BeeGFS seems to scale okay.

**Pull request merged**